sql/core/src/main/scala/org/apache/spark/sql/execution/RemoveRedundantProjects.scala

```diff
@@ -49,7 +49,11 @@ object RemoveRedundantProjects extends Rule[SparkPlan] {
     plan match {
       case p @ ProjectExec(_, child) =>
         if (isRedundant(p, child, requireOrdering) && canRemove(p, child)) {
-          removeProject(child, requireOrdering)
+          val newPlan = removeProject(child, requireOrdering)
+          // The `newPlan` should retain the logical plan link already. We call `setLogicalLink`
+          // here to make sure the `newPlan` sets the `LOGICAL_PLAN_TAG` tag.
+          newPlan.setLogicalLink(child.logicalLink.get)
+          newPlan
```

Comment:
Do you mean it has `LOGICAL_PLAN_INHERITED_TAG` but doesn't have `LOGICAL_PLAN_TAG`?

Reply:
Otherwise, retaining the logical plan link should mean `LOGICAL_PLAN_TAG` is set. Why we need to make sure it set again?

Reply:
> Do you mean it has LOGICAL_PLAN_INHERITED_TAG but doesn't have LOGICAL_PLAN_TAG?

Yes. `LOGICAL_PLAN_TAG` is kind of stronger than logical plan link (which means either LOGICAL_PLAN_TAG or LOGICAL_PLAN_INHERITED_TAG is set).